### PR TITLE
feat(connector): [HELCIM] Implement Cards for Helcim

### PR DESCRIPTION
**Pull request by deepanshu-iiitu**
## Type of Change

- [ ] Bugfix
- [x] New feature
- [ ] Enhancement
- [ ] Refactoring
- [ ] Dependency updates
- [ ] Documentation
- [ ] CI/CD

## Description

Helcim is a payment connector from North America.
Add Card payment support for Helcim.

Supported Flows:
- Authorize
- Capture
- PSync
- Void
- Refund
- RSync

### Additional Changes

- [ ] This PR modifies the API contract
- [ ] This PR modifies the database schema
- [x] This PR modifies application configuration/environment variables

## Motivation and Context

Add Card Payments for Payments and Refunds flow for new connector helcim.

## How did you test it?

Testing was carried out via Unit Tests

## Checklist

- [x] I formatted the code `cargo +nightly fmt --all`
- [x] I addressed lints thrown by `cargo clippy`
- [x] I reviewed the submitted code
- [x] I added unit tests for my changes where possible
- [ ] I added a [CHANGELOG](/CHANGELOG.md) entry if applicable

### CODE REVIEW

**Review by ArjunKarthik — CHANGES_REQUESTED** [review]

File: crates/router/src/connector/helcim.rs [hunk 1]
```diff
@@ -65,7 +67,13 @@ where
                 .into(),
         )];
         let mut api_key = self.get_auth_header(&req.connector_auth_type)?;
+
```
[the remark this answers is not in the document]
Comment (deepanshu-iiitu) [reply crates/router/src/connector/helcim.rs:68]: Done

File: crates/router/src/connector/helcim.rs [hunk 2]
```diff
@@ -350,15 +485,115 @@ impl ConnectorIntegration<api::Capture, types::PaymentsCaptureData, types::Payme
...
+        )];
+        let mut api_key = self.get_auth_header(&req.connector_auth_type)?;
+
+        let mut idempotency_key = vec![(
+            headers::IDEMPOTENCY_KEY.to_string(),
+            format!("{}V", req.payment_id).into_masked(),
+        )];
+        header.append(&mut api_key);
+        header.append(&mut idempotency_key);
+        Ok(header)
```
[the remark this answers is not in the document]
Comment (deepanshu-iiitu) [reply crates/router/src/connector/helcim.rs]: Done

File: crates/router/src/connector/helcim.rs [hunk 3]
```diff
@@ -350,15 +485,115 @@ impl ConnectorIntegration<api::Capture, types::PaymentsCaptureData, types::Payme
...
+        )];
+        let mut api_key = self.get_auth_header(&req.connector_auth_type)?;
+
+        let mut idempotency_key = vec![(
+            headers::IDEMPOTENCY_KEY.to_string(),
+            format!("{}_", req.request.refund_id).into_masked(),
+        )];
+        header.append(&mut api_key);
+        header.append(&mut idempotency_key);
+        Ok(header)
```
[the remark this answers is not in the document]
Comment (deepanshu-iiitu) [reply crates/router/src/connector/helcim.rs]: Done

File: crates/router/src/connector/helcim.rs [hunk 4]
```diff
@@ -106,15 +118,28 @@ impl ConnectorCommon for Helcim {
 
         Ok(ErrorResponse {
             status_code: res.status_code,
-            code: response.code,
-            message: response.message,
-            reason: response.reason,
+            code: response.errors.clone(),
+            message: response.errors,
+            reason: None,
```
[the remark this answers is not in the document]
Comment (deepanshu-iiitu) [reply crates/router/src/connector/helcim.rs]: Done

File: crates/router/src/connector/helcim.rs [hunk 5]
```diff
@@ -106,15 +118,28 @@ impl ConnectorCommon for Helcim {
...
-    //TODO: implement functions when support enabled
+    fn validate_capture_method(
+        &self,
+        capture_method: Option<enums::CaptureMethod>,
+    ) -> CustomResult<(), errors::ConnectorError> {
+        let capture_method = capture_method.unwrap_or_default();
+        match capture_method {
+            enums::CaptureMethod::Automatic
+            | enums::CaptureMethod::Manual
+            | enums::CaptureMethod::ManualMultiple => Ok(()),
```
[the remark this answers is not in the document]
Comment (deepanshu-iiitu) [reply crates/router/src/connector/helcim.rs]: Done

File: crates/router/src/connector/helcim/transformers.rs [hunk 6]
```diff
@@ -1,45 +1,181 @@
...
 pub struct HelcimPaymentsRequest {
-    amount: i64,
-    card: HelcimCard,
+    amount: f64,
+    currency: enums::Currency,
+    ip_address: Secret<String, IpAddress>,
+    card_data: HelcimCard,
+    billing_address: HelcimBillingAddress,
+    #[serde(skip_serializing_if = "Option::is_none")]
+    ecommerce: Option<bool>,
```
[the remark this answers is not in the document]
Comment (deepanshu-iiitu) [reply crates/router/src/connector/helcim/transformers.rs:69]: Added

File: crates/router/src/connector/helcim/transformers.rs [hunk 7]
```diff
@@ -1,45 +1,181 @@
...
-                    amount: item.request.amount,
-                    card,
+                    currency: item.request.currency,
+                    ip_address,
+                    card_data,
+                    billing_address,
+                    ecommerce: None,
+                })
+            }
+            _ => Err(errors::ConnectorError::NotImplemented("Payment methods".to_string()).into()),
```
[the remark this answers is not in the document]
Comment (deepanshu-iiitu) [reply crates/router/src/connector/helcim/transformers.rs]: Done

File: crates/router/src/connector/helcim/transformers.rs [hunk 8]
```diff
@@ -1,45 +1,181 @@
...
+                    name: req_address.get_full_name()?,
+                    street1: req_address.get_line1()?.to_owned(),
+                    postal_code: req_address.get_zip()?.to_owned(),
+                    street2: req_address.line2,
+                    city: req_address.city,
+                    email: item.request.email.clone(),
+                };
+                let ip_address = item
+                    .request
+                    .browser_info
```
[the remark this answers is not in the document]
Comment (deepanshu-iiitu) [reply crates/router/src/connector/helcim/transformers.rs]: Added

File: crates/router/src/connector/helcim/transformers.rs [hunk 9]
```diff
@@ -1,45 +1,181 @@
...
+                    .request
+                    .get_browser_info()?
+                    .get_ip_address()?;
+                Ok(Self {
+                    amount: item.amount.to_owned(),
+                    currency: item.router_data.request.currency,
+                    ip_address,
+                    card_data,
+                    billing_address,
+                    ecommerce: None,
```
[the remark this answers is not in the document]
Comment (deepanshu-iiitu) [reply crates/router/src/connector/helcim/transformers.rs]: Done

File: crates/router/src/connector/helcim/transformers.rs [hunk 10]
```diff
@@ -118,49 +539,73 @@ impl<F, T>
...
+}
 
-#[allow(dead_code)]
-#[derive(Debug, Serialize, Default, Deserialize, Clone)]
-pub enum RefundStatus {
-    Succeeded,
-    Failed,
-    #[default]
-    Processing,
+//TODO: Fill the struct with respective fields
```
[the remark this answers is not in the document]
Comment (deepanshu-iiitu) [reply crates/router/src/connector/helcim/transformers.rs]: Done

File: crates/router/src/connector/helcim/transformers.rs [hunk 11]
```diff
@@ -66,50 +202,335 @@ impl TryFrom<&types::ConnectorAuthType> for HelcimAuthType {
...
+impl TryFrom<&HelcimRouterData<&types::PaymentsCaptureRouterData>> for HelcimCaptureRequest {
+    type Error = error_stack::Report<errors::ConnectorError>;
+    fn try_from(
+        item: &HelcimRouterData<&types::PaymentsCaptureRouterData>,
+    ) -> Result<Self, Self::Error> {
+        let ip_address = item
+            .router_data
+            .request
+            .get_browser_info()?
+            .get_ip_address()?;
```
[the remark this answers is not in the document]
Comment (deepanshu-iiitu) [reply crates/router/src/connector/helcim/transformers.rs:479]: According to Helcim Support Team:
`If these transaction types will be merchant initiated, then passing the IP address of the original transaction is sufficient.`
Hence resolving this conversation.

File: crates/router/src/connector/helcim/transformers.rs [hunk 12]
```diff
@@ -66,50 +202,335 @@ impl TryFrom<&types::ConnectorAuthType> for HelcimAuthType {
...
+            HelcimPaymentsResponse,
+            types::PaymentsAuthorizeData,
+            types::PaymentsResponseData,
+        >,
     ) -> Result<Self, Self::Error> {
         Ok(Self {
-            status: enums::AttemptStatus::from(item.response.status),
             response: Ok(types::PaymentsResponseData::TransactionResponse {
-                resource_id: types::ResponseId::ConnectorTransactionId(item.response.id),
+                resource_id: types::ResponseId::ConnectorTransactionId(
```
[the remark this answers is not in the document]
Comment (deepanshu-iiitu) [reply crates/router/src/connector/helcim/transformers.rs]: While setting it to NoResponseId the capture transaction fails with the following error:
`{
    "error": {
        "type": "invalid_request",
        "message": "Resource ID does not exist in our records",
        "code": "HE_02"
    }
}`

Hence I am storing the `preauth_transaction_id` in `connector_transaction_id` as well as connector metadata during authorize call.
Later on while capture transaction I overwrite the `connector_transaction_id` with the `capture_id` received from connector.

File: crates/router/src/connector/helcim/transformers.rs [hunk 13]
```diff
@@ -118,49 +539,73 @@ impl<F, T>
...
 }
 
-impl<F> TryFrom<&types::RefundsRouterData<F>> for HelcimRefundRequest {
+impl<F> TryFrom<&HelcimRouterData<&types::RefundsRouterData<F>>> for HelcimRefundRequest {
     type Error = error_stack::Report<errors::ConnectorError>;
-    fn try_from(item: &types::RefundsRouterData<F>) -> Result<Self, Self::Error> {
+    fn try_from(
+        item: &HelcimRouterData<&types::RefundsRouterData<F>>,
+    ) -> Result<Self, Self::Error> {
+        let original_transaction_id = if item.router_data.request.connector_metadata.is_none() {
```
[the remark this answers is not in the document]
Comment (deepanshu-iiitu) [reply crates/router/src/connector/helcim/transformers.rs]: Done

File: crates/router/src/connector/helcim.rs [hunk 14]
```diff
@@ -106,15 +118,28 @@ impl ConnectorCommon for Helcim {
 
         Ok(ErrorResponse {
             status_code: res.status_code,
-            code: response.code,
-            message: response.message,
-            reason: response.reason,
+            code: response.errors.clone(),
```
[the remark this answers is not in the document]
Comment (deepanshu-iiitu) [reply crates/router/src/connector/helcim.rs]: Done

File: crates/router/src/connector/helcim/transformers.rs [hunk 15]
```diff
@@ -1,53 +1,242 @@
...
+        let card_data = HelcimCard {
+            card_expiry: req_card.get_card_expiry_month_year_2_digit_with_delimiter("".to_string()),
+            card_number: req_card.card_number.clone(),
+            card_c_v_v: req_card.card_cvc.clone(),
+        };
+        let req_address = item
+            .get_billing()?
+            .to_owned()
+            .address
+            .ok_or_else(utils::missing_field_err("billing.address"))?;
```
Comment (ArjunKarthik) [inline crates/router/src/connector/helcim/transformers.rs:107]: let req_address = item.get_billing_address()?;

File: crates/router/src/connector/helcim/transformers.rs (hunk 12 above)
[the remark this answers is not in the document]
Comment (ArjunKarthik) [reply crates/router/src/connector/helcim/transformers.rs]: In which step you are getting the above mentioned error? We should not rotate the connector_transaction_id. It is only primary id to identify the transaction at connector side

[the remark this answers is not in the document]
Comment (ArjunKarthik) [reply crates/router/src/connector/helcim/transformers.rs]: Let me know, if incase you need discussion on this

**Review by ArjunKarthik — CHANGES_REQUESTED** [review]

File: crates/router/src/connector/helcim/transformers.rs (hunk 15 above)
Comment (deepanshu-iiitu) [reply crates/router/src/connector/helcim/transformers.rs]: Done

File: crates/router/src/connector/helcim.rs [hunk 16]
```diff
@@ -131,6 +160,72 @@ impl ConnectorIntegration<api::AccessTokenAuth, types::AccessTokenRequestData, t
 impl ConnectorIntegration<api::Verify, types::VerifyRequestData, types::PaymentsResponseData>
     for Helcim
 {
```
Comment (SamraatBansal) [inline crates/router/src/connector/helcim.rs:162]: Merge with main to update the flow

File: crates/router/src/connector/helcim.rs [hunk 17]
```diff
@@ -273,6 +394,12 @@ impl ConnectorIntegration<api::PSync, types::PaymentsSyncData, types::PaymentsRe
     ) -> CustomResult<ErrorResponse, errors::ConnectorError> {
         self.build_error_response(res)
     }
+
+    // fn get_multiple_capture_sync_method(
+    //     &self,
+    // ) -> CustomResult<services::CaptureSyncMethod, errors::ConnectorError> {
+    //     Ok(services::CaptureSyncMethod::Individual)
+    // }
```
Comment (SamraatBansal) [inline crates/router/src/connector/helcim.rs:402]: Is this required?

**Review by SamraatBansal — CHANGES_REQUESTED** [review]

File: crates/router/src/connector/helcim/transformers.rs (hunk 12 above)
[the remark this answers is not in the document]
Comment (deepanshu-iiitu) [reply crates/router/src/connector/helcim/transformers.rs]: Implemented

File: crates/router/src/connector/helcim.rs (hunk 16 above)
Comment (deepanshu-iiitu) [reply crates/router/src/connector/helcim.rs]: Done

File: crates/router/src/connector/helcim.rs (hunk 17 above)
Comment (deepanshu-iiitu) [reply crates/router/src/connector/helcim.rs:414]: Yes.

**Review by ArjunKarthik — APPROVED** [review]

**Review by SamraatBansal — APPROVED** [review]

File: crates/common_enums/src/transformers.rs [hunk 18]
```diff
@@ -10,6 +10,261 @@ impl Display for NumericCountryCodeParseError {
     }
 }
 
+impl CountryAlpha2 {
+    pub const fn from_alpha2_to_alpha3(code: Self) -> CountryAlpha3 {
```
Comment (SanchithHegde) [inline crates/common_enums/src/transformers.rs:14]: I'd prefer naming this just `to_alpha3`.

**Review by SanchithHegde — APPROVED** [review]
Other than that, looks good to me!

**Pull request merged**